core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.stream.ByteBufsInputStream;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    private final StreamMessage<T> source;
+    private final EventExecutor executor;
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    private boolean subscribed;
+    private boolean closed;
+
+    StreamMessageInputStream(StreamMessage<T> source,
+                             Function<? super T, ? extends HttpData> httpDataConverter,
+                             EventExecutor executor) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        requireNonNull(executor, "executor");
+        this.source = source;
+        this.executor = executor;
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(in -> {
+            try {
+                return in.read();
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return read(in -> {
+            try {
+                return in.read(b, off, len);
```

Comment:
The code savings probably isn't worth using a lambda here, notably these parameters will need to be captured into an allocating lambda I think

Reply:
```java
    @Override
    public int read() throws IOException {
        request();
        return byteBufsInputStream().read();
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        request();
        return byteBufsInputStream().read(b, off, len);
    }

    private void request() throws IOException {
        ensureOpen();
        ensureSubscribed();
        subscriber.request();
    }
```
> The code savings probably isn't worth using a lambda here

aha.. then above code would be better than now? 🤔

Reply:
JFYI, you can test the difference of reference IDs of returned `Function`s like:
```java
@Test
void test() {
    for (int i = 0; i < 10; i++) {
        System.err.println(local(i, i + 1));
        System.err.println(nonLocal());
    }
}

Function<Integer, Integer> local(int x, int y) {
    return num ->  num + x + y;
}

Function<Integer, Integer> nonLocal() {
    return num ->  num + 1;
}
```